### PR TITLE
Support for data element order and filter on /event endpoint

### PR DESCRIPTION
**Pull request by Bekkalizer**

### CODE REVIEW

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/event/EventController.java [hunk 1]
```diff
@@ -1065,12 +1103,32 @@ protected boolean hasHref( List<String> fields )
...
-
             return Arrays.asList( order.split( "," ) );
         }
 
         return null;
     }
+    
+    private List<String> getGridOrderParams( String order, Map<String,String> dataElementOrders )
+    {
+        List<String> dataElementOrderList = new ArrayList<String>();
```
Comment (huzvak) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/event/EventController.java]: <>

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/event/EventController.java [hunk 2]
```diff
@@ -1065,12 +1103,32 @@ protected boolean hasHref( List<String> fields )
...
+        
+        if ( order != null && !StringUtils.isEmpty( order ) && dataElementOrders != null && dataElementOrders.size() > 0 )
```
Comment (huzvak) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/event/EventController.java]: If you use StringUtils from Spring package, then you don't have to do null check yourself

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/JdbcEventStore.java [hunk 3]
```diff
@@ -602,6 +613,75 @@ private String getEventSelectQuery( EventSearchParams params, List<OrganisationU
...
+            {
+                final String col = statementBuilder.columnQuote( item.getItemId() );
+                sql += ( item.hasFilter() ? "inner" : "left" ) + " join trackedentitydatavalue as " + col + " " + "on " + col
+                    + ".programstageinstanceid = psi.programstageinstanceid " + "and " + col + ".dataelementid = "
+                    + item.getItem().getId() + " ";
+                for ( QueryFilter filter : item.getFilters() )
+                {
+                    final String encodedFilter = statementBuilder.encode( filter.getFilter(), false );
+    
+                    final String queryCol = item.isNumeric() ? " CAST( " + (col + ".value AS NUMERIC)")
```
[the remark this answers is not in the document]
Comment (Bekkalizer) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/JdbcEventStore.java]: encodedFilter contains filter being looped, so this has to stay within the loop - maybe I'll keep querycol together for readability. Also - many items will not have a filter, and anything outside the loop will be run for those with no filters...

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/JdbcEventStore.java [hunk 4]
```diff
@@ -959,13 +1038,37 @@ private String getGridOrderQuery( EventSearchParams params )
...
-            ArrayList<String> orderFields = new ArrayList<String>();
-
-            for ( Order order : orders )
+            for ( String order : params.getGridOrders() )
+            {
+                String[] prop = order.split( ":" );
+    
+                if ( prop.length == 2 && ( prop[1].equals( "desc" ) || prop[1].equals( "asc" ) ) )
+                {
+                    Iterator<QueryItem> itermIterator = params.getDataElements().iterator();
```
[the remark this answers is not in the document]
Comment (Bekkalizer) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/JdbcEventStore.java]: Typo - did you mean the missing spaces in line 1051?
It was a good idea to use a for-loop instead, I copied this strategy from somewhere and didnt think more of it.

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/JdbcEventStore.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bekkalizer) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/event/JdbcEventStore.java]: encodedFilter contains filter being looped, so this has to stay within the loop - maybe I'll keep querycol together for readability. Also - many items will not have a filter, and anything outside the loop will be run for those with no filters...

**Pull request merged**